### DISCUSSION TITLE
Filter blocks: being the extender ourselves

### DISCUSSION THREAD

**Discussion by dinhtungdu** [woocommerce/woocommerce#44944] [WooCommerce, Blocks, and Site Editing]
As you might, or might not, know we’re working on a new set of filter blocks with better performance and experience. Not only user experience, we want to improve the developer experience as well. In this post, I’m explaining the structure I’m experimenting with to make filter blocks extensible.

## The goal

The way I’m defining extensibility in the context of filter blocks is the ability to change/control the visual of filter blocks in both the editor and front end. In other words, WC Blocks provides the base block with a default representation, and developers can change that representation to fit their design needs, without recreating complex logic like fetching collection data and passing the context.
  
![image](https://github.com/woocommerce/woocommerce-blocks/assets/5423135/0373d0d3-f09d-4991-809b-12463a3022de)

Take the Price Filter block as an example. Currently, WC Blocks ships it with a Price Slider. Developers can’t change the look and feel of that block. What if the site doesn’t want a slider but a range selector instead? Developers shouldn’t have to recreate/mimic the whole price filter block to just change the visual look.

![image](https://github.com/woocommerce/woocommerce-blocks/assets/5423135/6af294d3-65a4-4ee6-a83a-fc03b5af9d5e)

At the end of the day, we only need the min and maximum price to narrow down the product grid (or list). So if we can provide a set of shared data (let’s call it state) and actions, we can have different representations for the same block. For a price filter block, they could be:

- Actions to set the min and maximum prices.
- Actions to submit/apply filter data (e.g.: the action attached to the submit button)
- State for current min/max prices.
- State for the actual price range of current qualified products in the query.

  

Applying the same thinking to other filter blocks like the Attribute filter, the shared actions and state can be:

- Actions to set the currently selected attributes.
- Actions to submit/apply filter data.
- State for currently available attributes.
- State for the currently selected attributes.

## The proposal

With the above mindset in mind, and the new power that Interactivity API gives us, I want to propose a structure for new filter blocks: Create filter blocks as parent blocks that provide state and actions, and use inner blocks as representation (or view) that consume those state and actions. Doing so has several advantages:

- Separation of concern:
    - Take the price filter example again: currently, the price slider has two settings: show/hide the price input field and display the input inline with the slider. Those settings don’t make sense for a ‘Price Range’ filter block. Using inner blocks, we can keep the relevant settings/options only where they need to be.
    - Or take one of the most frequently asked questions IMO: how can we display the actual color for an attribute? With an inner block, we can add inspector settings, only for that block, to enable merchants to pick the desired color for each attribute, then use that setting in the preview and frontend output.
- Developer friendly: To create a new view, all developers need to do is create a block. We don’t introduce any new terminology or registration process, every view is just another Gutenberg block.
- Optimized assets loading out of the box: by keeping the relevant assets within the ‘view’ block, we don’t need to load bloated/unused style/script files.
- Using block registry, we can query available inner blocks of a specific filter block. Then we can create a view switcher/selector for each filter block to let the merchant switch between them. Again, developers need to do nothing but register the inner blocks. We can even utilize the block preview to have a nice-looking and intuitive view switcher/selector.
- Extensible by design: The default view should be also an inner block, which ensures extensibility at some levels.

## The POC

I created [#11139](https://github.com/woocommerce/woocommerce-blocks/pull/11139 "https://github.com/woocommerce/woocommerce-blocks/pull/11139") to demonstrate my idea explained above:

- The `Collection Price Filter`  is a parent block:
    - It provides Interactivity API states and actions. Note that it doesn’t provide the `rangeStyle`  anymore.
    - It passes the common states into inner blocks via context ( `filterData` ).
    - Also note that it doesn’t have `showInputFields`  and  `inlineInput`  anymore.
- The `Collection Price Filter Slider`  is the inner block of the `Collection Price Filter`  block:
    - It consumes the `filterData`  context to render the preview in the editor as well as the block output in the front end.
    - It provides the `rangeStyle`  state, which is only relevant to itself, the slider.
    - The `showInputFields`  and  `inlineInput`  attributes now belong to this inner block. Again, they’re only relevant to a price slider.

If developers want a different visual look, like the price range above, they can create a block. We don’t introduce any new terminology or helper for developers to learn, they just create blocks as they’re already doing.

I created a short demo video to demo my idea:

[![](https://cdn.loom.com/sessions/thumbnails/a8f2826248e245dc95e4351079570a78-with-play.gif)](https://www.loom.com/share/a8f2826248e245dc95e4351079570a78)

[](https://www.loom.com/share/a8f2826248e245dc95e4351079570a78?sid=f04cfa70-8f80-4f5a-b324-2671cc39c1af)[Using Inner Block for the New Collection Filter Block - Watch Video](https://www.loom.com/share/a8f2826248e245dc95e4351079570a78)  

The plugin I created in the demo: [https://github.com/dinhtungdu/price-filter-range](https://github.com/dinhtungdu/price-filter-range)

Let me know what you think about this, should we go ahead with it? I can see this makes filter blocks more complex, but I consider it as the trade-off for extensibility.

**Comment by dinhtungdu** [comment]
https://github.com/woocommerce/woocommerce-blocks/blob/f73d7013c9cf6e59dbc4bd2c45680909021ba45c/assets/js/blocks/collection-filters/inner-blocks/price-filter-slider/frontend.ts#L16-L28

This is subjected to change when the new [New Interactivity store() API proposal](https://github.com/WordPress/gutenberg/discussions/53586) by @luisherranz is approved and implemented.

**Comment by luisherranz** [comment]
This is very interesting. Thanks for sharing, Tung 🙂

Could you please list the things that would need to be shared across all the "view blocks"? Is a parent block the best way to share them, or are there other options to consider?

**Reply by dinhtungdu** [reply]
> Could you please list the things that would need to be shared across all the "view blocks"?

@luisherranz Each filter block with have a different specification, so their "view blocks" will consume different sets of states/actions depending on their parent filter blocks. But generally, in the context of Interactivity API, "view blocks" should be able to access all states and actions of parent filter blocks.

> Is a parent block the best way to share them, or are there other options to consider?

I choose parent block because it's the most optimal solution for various aspects:
- Performance and assets loading: only relevant assets are loaded by default, without any optimization from our side.
- Developer experience: To create a new view, developers only need to create blocks. They don't have to learn any new set of APIs (and we also don't need to maintain those APIs).
- Reusability: We can use existing Block APIs to build the view switcher.

I was considering other approach until @mtias suggested using inner blocks:
- Using [editor.BlockEdit](https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#editor-blockedit) in combination with `<Slot>`/`<Fill>` to let develops hook into to the editor.
- Using a custom WordPress filter to filter the block output.
But by doing so, we're introducing new things, and it's harder/more complicated compared to registering a block from the developer's POV.

I haven't think of other alternatives, if you have any in your mind, please share with us!

**Reply by luisherranz** [reply]
First, I'd like to know the list of things that would need to be shared across all the view blocks of the price filter.

For example, hypothetically:

- In the client
  - A "Store part" that contains some actions/state:
    - `actions.updatePrice(min, max)`
    - `state.minPrice`
    - `state.maxPrice`
  - An element wrapper that contains some context:
    - `<div data-wp-context='{...}'>`
- In the server
  - The logic to retrieve the min/max prices:
    - `$min_price`
    - `$max_price`

In other words, what actions/state/context/server-logic… are you abstracting in the parent block?

It would also be helpful to know the things that will differ and need to be provided by the new implementation, like for example the markup, the styles (CSS), additional state/actions, etc.

**Reply by dinhtungdu** [reply]
For price filter block:

- Client: the parent block in my [example](https://github.com/woocommerce/woocommerce-blocks/pull/11139) is providing these actions/state that are meant to be consumed by view blocks:
  - Actions:
    - `setMinPrice`
    - `setMaxPrice`
    - `updateProducts`
  - State:
    - `minPrice`
    - `maxPrice`
    - `minRange`
    - `maxRange`
    - `formattedMinPrice`
    - `formattedMaxPrice`
  - Element: with the current version of WC Interactivity API, I'm current using `wc_store()` (PHP) and `store()` (JS) to prepare the store so the parent block doesn't need to declare context. But when the new proposal is implemented, we must declare it, as you suggested.
- Server:
  - We passed down the state to inner blocks via context (named `filterData`). So inner blocks don't need to do anything to retrieve the data, just get them from the context: https://github.com/woocommerce/woocommerce-blocks/blob/f73d7013c9cf6e59dbc4bd2c45680909021ba45c/src/BlockTypes/CollectionPriceFilter.php#L38-L51

> It would also be helpful to know the things that will differ and need to be provided by the new implementation, like for example the markup, the styles (CSS), additional state/actions, etc.

- Each view block has its markup/style/asset. The parent filter block doesn't need to care about them.
- Each view can have additional state actions if needed, for example, the slider view block needs a derived `rangeStyle` state to calculate thumb position, which is [defined in the slider view block](https://github.com/woocommerce/woocommerce-blocks/blob/f73d7013c9cf6e59dbc4bd2c45680909021ba45c/assets/js/blocks/collection-filters/inner-blocks/price-filter-slider/frontend.ts#L16-L28), not the parent one.

I feel that I do not fully understand what you mean 😅, happy to jump on a call to chat about this.

**Reply by luisherranz** [reply]
Thanks, Tung. I think this is an excellent exercise, but rather than jumping immediately to a specific solution, I would like to analyze the requirements in order to weigh the different extensibility options.

---

Let's first analyze the value that a parent "Collection Price Filter" block would add to a user. This is my reasoning:

- At first, it may seem nice to have a selector that shows all the "view price filter blocks" in one place.
- But we need to take into account that users are not educated to search for blocks in the inspector panel. They will most likely search for their new price filter block in the inserter.
- In the inserter, the new price filter block will only be available when the user tries to insert it as a child of the Collection Price filter.
- If they try to insert it as a child of the Collection Filters block, it won't appear. That's problematic because imagine a user has deleted the Collection Price Filter block. It's not intuitive that they need to add the Collection Price Filter block back.
- The Collection Price Filter block does not provide any meaningful settings (other than choosing the "view block").
- The majority of users (+90%) will probably never use a different view price filter block, and yet they will see an "unnecessary" additional block on their sites.

So, I think one could argue that the Collection Price Filter is not adding any real value to the user. Actually, it could add a bit of confusion.

Any opinions?

---

Now, let's analyze the requirements of the view blocks:

- **Context**: They need to access some context local to the current Products Collection block (probably using `data-wp-context`).
- **Store part**: They need to access some actions and derived state in the client (JS).
- **Data**: They need to access some filter data in the server (PHP).

Are they correct? Anything else?

---

Now, what options do we have?

1. Use a parent Collection Price Filter block (current exploration):

   - **Context**: Put it in the Collection Price Filter block wrapper.
   - **Store part**: Enqueue it as the Collection Price Filter view script.
   - **Data**: Pass it down using Block context.

   Instructions for the block developer:

   - Create a new block.
   - Choose the Collection Price Filter block as an ancestor.
   - Use the Block context to get the data.

   Pros and cons:

   - **Con**: An extra block might be confusing for the user.
   - **Pro**: The block developer doesn't need to take care of enqueuing the view script.
   - **Pro**: The block developer doesn't need to take care of adding `data-wp-context` themselves.

2. Do something similar, but use the Collection Filter block to pass things down:

   - **Context**: Put it in the Collection Filter block wrapper.
   - **Store part**: Only register it, but don't enqueue it by default. _We don't want to enqueue things that might not be necessary._
   - **Data**: Pass it down using Block context.

   Instructions for the block developer:

   - Create a new block.
   - Choose the Collection Filter block as an ancestor.
   - Use the Block context to get the data.
   - Add the store part script as a dependency (or add it to the view scripts array).

   Pros and cons:

   - **Pro**: No extra block.
   - **Con**: The block developer needs to take care of enqueuing the store part script.
   - **Pro**: The block developer doesn't need to take care of adding `data-wp-context` themselves.
   - **Con**: The Collection Filter will process and pass down the `data-wp-context` and Block context unnecessarily if there is no price filter block.

   Questions:

   - Could the Collection Filter block inspect the presence of a price filter block? _It doesn't seem obvious to me at this moment._

3. Don't use any parent block at all:

   - **Context**: Expose a function to retrieve it (like `wc_collection_price_filter_context()`).
   - **Store part**: Only register it, but don't enqueue it by default.
   - **Data**: Expose a function to retrieve it (like `wc_collection_price_filter_data()`).

   Instructions for the block developer:

   - Create a new block.
   - Choose the Collection Filter block as an ancestor.
   - Get some sort of Collection ID from the Block context.
   - Use the `wc_collection_price_filter_context( $id )` function to get the context and add it manually to the block wrapper using `data-wp-context`.
   - Add the store part script as a dependency (or add it to the view scripts array).
   - Use the `wc_collection_price_filter_data( $id )` function to get the data.

   Pros and cons:

   - **Pro**: No extra block.
   - **Con**: The block developer needs to take care of enqueuing the store part script.
   - **Con**: The block developer needs to take care of using `data-wp-context` themselves.
   - **Pro**: The Collection Filter will not process and pass down the `data-wp-context` and Block context unnecessarily if there is no price filter block.
   
   Questions: 

   - How complex is it to expose the necessary logic using functions that only get some Collection ID?

Questions:

- Any other options or ideas?
- What do you think about the cons? Are any deal-breakers?
- Are the benefits of using an additional parent block justified?

---

In general, the UX is much more important than the DX. Take into account that there will be thousands of users using the price filter blocks, while there will be only a bunch of developers creating new price filter blocks. So we should usually optimize for the former.

Again, I'm not especially opposed to the inner block pattern. Actually, I think this is a great and necessary exploration, and for that reason, we should analyze all the possible options 🙂

**Reply by dinhtungdu** [reply]
@luisherranz Excellent reply and questions! Thank you so much for your time and effort putting into this.

> - But we need to take into account that users are not educated to search for blocks in the inspector panel. They will most likely search for their new price filter block in the inserter.

Agreed.

> - In the inserter, the new price filter block will only be available when the user tries to insert it as a child of the Collection Price filter.
> - If they try to insert it as a child of the Collection Filters block, it won't appear. That's problematic because imagine a user has deleted the Collection Price Filter block. It's not intuitive that they need to add the Collection Price Filter block back.
> - The majority of users (+90%) will probably never use a different view price filter block, and yet they will see an "unnecessary" additional block on their sites.

This is not the case with my current idea: The view blocks shouldn't be available in the inserter. They should be manipulated by the 'view switcher' only. So in the inserter, there should be only one 'Collection Price Filter' block. In my POC PR, I also disabled the appender for the 'Collection Price Filter' block, so the visibility of view blocks should be minimal.

> - The Collection Price Filter block does not provide any meaningful settings (other than choosing the "view block").

I agree, but I want to follow the same strategy Product Gallery block is doing: exposing the view block settings into its parent block. So any view block settings should appear under the view switcher.

So with all of those additions and adjustments, I think a 'Collection Price Filter' block is less confusing and provides more value to end users.

* * *

> Now, let's analyze the requirements of the view blocks:  
> 
> - Context: They need to access some context local to the current Products Collection block (probably using data-wp-context).
> - Store part: They need to access some actions and derived state in the client (JS).
> - Data: They need to access some filter data in the server (PHP).
> 
> Are they correct? Anything else?

Correct. About context, the 'Collection Filters' block will be the one that provide context. It will prepare and pass the context down to the inner filter block (on both JS and PHP sides).

* * *

> Could the Collection Filter block inspect the presence of a price filter block? It doesn't seem obvious to me at this moment.

Yes. We're doing so [already](https://github.com/woocommerce/woocommerce-blocks/blob/92076d169f16eeb30175eaf92ad95eaa41f07ab2/src/BlockTypes/CollectionFilters.php#L151 "https://github.com/woocommerce/woocommerce-blocks/blob/92076d169f16eeb30175eaf92ad95eaa41f07ab2/src/BlockTypes/CollectionFilters.php#L151"). So the option 2 looks quite promising to me. But doing everything in `Collection Filters`  sounds like a code smell to me.

> - How complex is it to expose the necessary logic using functions that only get some Collection ID?

I would say the complexity level is low for that type of logic.

> - Any other options or ideas?

I don’t have any.

> - What do you think about the cons? Are any deal-breakers?

- For option 1, with my additional info above, do you it solves the con you mentioned (**Con**: An extra block might be confusing for the user.).
- Most of the cons aren’t deal-breakers, except these:

    - The block developer needs to take care of using `data-wp-context` themselves.

    - **Con**: The Collection Filter will process and pass down the `data-wp-context` and Block context unnecessarily if there is no price filter block.

 One of the benefits and reason for using ‘Collection Filters’ is to prepare the collection data for all inner blocks. We detect inner filter blocks and prepare the collection data for all inner blocks in one place using only one internal API request. By doing so, we ensure the performance of filter blocks and keep the logic of filter blocks simple (because they only need to consume the context passed from the `Collection Filters`  block).  

> - Are the benefits of using an additional parent block justified?

  

To me, my biased opinion is yes:

- As I mentioned above, if we disable the inserter of view blocks and appender of filter blocks, and show the view block settings in the filter block sidebar, then using option 1 is much less confusing. Most user may not change the view, but they also don’t need to know about other views if exists.
- Option 2 is also worth exploring. We can detect inner filter blocks, so the performance will be the same as option 1. We also can enqueue only necessary store data, in only one place (which is a gain). The only thing I worry about is we may put too much logic inside `Collection Filters`, which is a smell to me. Otherwise, I really like this option, as it simplifies both block and source code structure. Filter blocks are now all ‘view blocks’, which make theme even easier to write.
- For option 3, I don’t think we should go with that path, as it makes writing filter blocks harder. We’re also exposing new functions, which we have to maintain in the future.

**Reply by kimmenbert** [reply]
> In general, the UX is much more important than the DX. Take into account that there will be thousands of users using the price filter blocks, while there will be only a bunch of developers creating new price filter blocks. So we should usually optimize for the former.

I understand this is important. But, I do think that whenever facing this, you should be thinking of ways to enable agencies / developers to "break" the UX when _needed_. Where WooCommerce out of the box is UX first, but developers could make what they need. Thinking out loud something like `hasPriceFilterModifyingBlock` where devs would register a block, triggering this to "true" somehow, and more options to be available, eg. price filter block now support inner block.

I understand that this opens the option for some random plugin author to make a plugin that ruins UX without caring. But for an agency this would be handled with care within specific projects.

**Reply by dinhtungdu** [reply]
> Option 2 is also worth exploring. We can detect inner filter blocks, so the performance will be the same as option 1. We also can enqueue only necessary store data, in only one place (which is a gain). The only thing I worry about is we may put too much logic inside Collection Filters, which is a smell to me. Otherwise, I really like this option, as it simplifies both block and source code structure. Filter blocks are now all ‘view blocks’, which make theme even easier to write.

For this, we need to expose `useCollectionData` hooks publicly so developers can build the editor preview. We also need to expose the logic to get context from product collection too. Developers now don't write 'pure' view blocks anymore, as they need to do some logic in the editor.

**Reply by luisherranz** [reply]
Can you show examples of the view block Edit components from both option 1 and option 2 to understand the difference?

**Reply by dinhtungdu** [reply]
@luisherranz 

- For option 1, you can check my working example here: https://github.com/woocommerce/woocommerce-blocks/pull/11139/files.
The Edit component of the view block is here: https://github.com/woocommerce/woocommerce-blocks/blob/try/inner-filter-block/assets/js/blocks/collection-filters/inner-blocks/price-filter-slider/edit.tsx
You can see that it only consumes the context passed down by parent block.

- For option 2, I haven't finished the working example yet, but the edit component of the view block should look like this: https://github.com/woocommerce/woocommerce-blocks/blob/try/new-price-filter/assets/js/blocks/collection-filters/inner-blocks/price-filter/edit.tsx
Notice that it use the `useCollectionData` hook. It will also need to use another shared util to transform the Product Collection query into the Store API/ ProductCollectionData route parameters.

In the editor, we need to use `useCollectionData` hooks to fetch the collection data when the block is rendered. I thought about fetching all available collection data but quickly discarded it because we're fetching unnecessary data, and it may hurt the performance on larger stores with many products.

**Reply by luisherranz** [reply]
Can you share this link again? It's now returning a 404.

https://github.com/woocommerce/woocommerce-blocks/blob/try/new-price-filter/assets/js/blocks/collection-filters/inner-blocks/price-filter/edit.tsx

(by the way, it's always safer to [share the permalinks](https://docs.github.com/en/repositories/working-with-files/using-files/getting-permanent-links-to-files#press-y-to-permalink-to-a-file-in-a-specific-commit) instead 🙂)

>  We can detect inner filter blocks, so the performance will be the same as option 1

How are you planning to do this? For example, imagine I have replaced the official "Price Filter block" with "My Custom Filter block". How does the Collection Filter block know that "My Custom Filter block" needs the price data?

**Reply by dinhtungdu** [reply]
@luisherranz I fixed the link, the branch was auto deleted after merge, I restored the branch and the link is working now.

> How are you planning to do this? For example, imagine I have replaced the official "Price Filter block" with "My Custom Filter block". How does the Collection Filter block know that "My Custom Filter block" needs the price data?

View blocks need to "tell" `Collection Filters` about what they needs. There are some options to do achieve that: using PHP filter or use block metadata (attributes, block support, block category). PHP filter maybe the most straight forward option to work on both frontend and editor.

**Reply by dinhtungdu** [reply]
I'm working on a POC to fully demonstrate my idea.

**Reply by luisherranz** [reply]
Block metadata sounds like a good option to explore as it can be accessed from both PHP and JS, although I don't know the general Gutenberg guidelines about including arbitrary data. We can ask 🙂

**Reply by dinhtungdu** [reply]
@luisherranz I updated my POC PR (https://github.com/woocommerce/woocommerce-blocks/pull/11191/files) for option 2:
- It's fully working and a complete solution.
- I use an attribute called `filterType` belong to view block to indicate the type of filter. Collection Filters block also use that information to build the required collection data query, on both frontend and editor.
- Collection Filters block provides a new context named `filterData` which the Interactivity API state data. This data is processed from collection data fetched from Store API.

To me, I think this is the most simple and optimal approach (in terms of both UX and DX) among ones discussed in this thread. What do you think?

**Reply by luisherranz** [reply]
Catching up here 🙂

I'd like to get a second opinion about using an attribute to store a block property. But other than that, I agree this approach seems clean and simple. Great work!

**Comment by imanish003** [comment]
Hi @dinhtungdu,

Fantastic proposal! I'm particularly enthusiastic about the focus on both developer and user experience. I do have a few considerations:

**Inner Block Placement:** We need to ensure that these inner blocks are restricted to appearing only within their designated parent blocks. It's crucial that we clearly communicate this constraint to third-party developers (3PDs).

**Security and Future Compatibility:** It's important to proactively address any potential security vulnerabilities. Additionally, as the Interactivity API evolves, there may be new features like namespaces and permission levels for stores. We should ensure our current work aligns well with those potential future developments.

**User Experience for Inner Blocks:** I noticed that the styles dropdown in the demo could be improved for a better user experience.

Overall, I wholeheartedly agree that this approach has the potential to set a positive precedent for our future blocks as well 🚀

**Reply by dinhtungdu** [reply]
> Inner Block Placement: We need to ensure that these inner blocks are restricted to appearing only within their designated parent blocks. It's crucial that we clearly communicate this constraint to third-party developers (3PDs).

@imanish003 the inner blocks must declare the ancestor/parent block to a filter block to be appear in the view switcher, so I won't worry much about that. That's also the only constraint developers need to follow when registering inner blocks. And yes, we really need clear communication here.

> Security and Future Compatibility: It's important to proactively address any potential security vulnerabilities. Additionally, as the Interactivity API evolves, there may be new features like namespaces and permission levels for stores. We should ensure our current work aligns well with those potential future developments.

Yes, I acknowledge the https://github.com/WordPress/gutenberg/discussions/53586. We will work with Frontity folks to ensure we ship the proper Interactivity blocks.

> User Experience for Inner Blocks: I noticed that the styles dropdown in the demo could be improved for a better user experience.

Yep, as I mentioned in the post, I want to use block preview to render a visual view switcher. I use a dropdown in the demo just for quick POC:
> We can even utilize the block preview to have a nice-looking and intuitive view switcher/selector.

**Reply by imanish003** [reply]
Sounds great! Thanks 🚀

**Comment by luisherranz** [comment]
@dinhtungdu can you do an update here on the latest work on this pattern? I see there has been some related work on [11476](https://github.com/woocommerce/woocommerce-blocks/pull/11476), although I'm not sure if that is the latest PR.

My impression is that you're moving from passing down the processed data via context to abstracting that processing into separate functions that view blocks can call, but are more flexible because they can pass their own arguments (i.e., terms for the Attribute filter). Is that correct?

**Reply by dinhtungdu** [reply]
@luisherranz for now, we're putting the work on the extensibility side on hold to reach the feature parity with current blocks faster ( creating a equivalent set of filter blocks compared to the current ones), but we'll revisit this topic later after that.

Also, we tried the option 2 as we discussed [here](https://github.com/woocommerce/woocommerce-blocks/discussions/11141#discussioncomment-7288078) but we abandoned it (see https://github.com/woocommerce/woocommerce-blocks/issues/11306#issuecomment-1782220553). I'm looking to revisit the original idea in the next phase as I still see as the best approach overall.
